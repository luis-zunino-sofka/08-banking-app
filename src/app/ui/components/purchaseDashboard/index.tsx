import { Toaster } from "react-hot-toast";
import { usePurchase } from "@core/hooks";
import { decryptAES } from "@core/utils";
import { Button } from "../button";
import "../../styles/styles.scss";

export const PurchaseDashboard = () => {
  const { isLoading, accounts, errors, register, handleSubmit, onSubmit } =
    usePurchase();

  return (
    <div>
      <form
        id="purchase-form"
        className="form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <div>
          <label htmlFor="purchase-type">Selecciona el tipo de compra:</label>
          <select
            id="purchase-type"
            {...register("purchaseType", {
              required: "Por favor selecciona un tipo de compra",
            })}
          >
            <option value="">Elige el tipo de compra</option>
            <option value="physical">Establecimiento físico (Sin costo)</option>
            <option value="online">En línea (Costo de seguro de $5 USD)</option>
          </select>
          {errors.purchaseType && (
            <span className="error-message">{errors.purchaseType.message}</span>
          )}
        </div>
        <div>
          <label htmlFor="account-number-to-deposit">
            Selecciona la cuenta a depositar:
          </label>
          <select
            id="account-number-to-deposit"
            {...register("accountId", {
              required: "La cuenta bancaria a depositar es obligatorio",
            })}
          >
            <option value="">Elige una cuenta</option>
            {accounts.map((a) => (
              <option key={a.accountId} value={a.accountId}>
                Cuenta: {decryptAES(a.encryptedNumber)} - Monto: {a.amount}
              </option>
            ))}
          </select>

          {errors.accountId && (
            <span className="error-message">{errors.accountId.message}</span>
          )}
        </div>
        <div>
          <label htmlFor="purchase-amount">
            Ingresa el monto a depositar:
            <input
              type="number"
              id="purchase-amount"
              placeholder="Monto de la compra en $"
              {...register("purchaseAmount", {
                required: "El monto es obligatorio",
                min: {
                  value: 1,
                  message: "El monto debe ser mayor que $0",
                },
              })}
            />
          </label>
          {errors.purchaseAmount && (
            <span className="error-message">
              {errors.purchaseAmount.message}
            </span>
          )}
        </div>
        <Button label="Comprar" isLoading={isLoading} />
      </form>

      <Toaster />
    </div>
  );
};
