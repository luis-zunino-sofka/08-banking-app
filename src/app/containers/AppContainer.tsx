import { useNavbar } from "@core/hooks";
import { useAccountContext, useAppContext } from "@core/state";
import { Navbar } from "@ui/components/navbar";
import { AppLayout } from "@ui/layouts";
import React from "react";
import { Toaster } from "react-hot-toast";

export const AppContainer = ({ children }: { children: React.ReactNode }) => {
  const { state } = useAppContext();
  const { refetchAccounts } = useAccountContext();
  const { navItems, handleLogout } = useNavbar();

  return (
    <AppLayout>
      <Navbar
        state={state}
        handleLogout={handleLogout}
        refetchAccounts={refetchAccounts}
        navItems={navItems}
      />
      {children}
      <Toaster />
    </AppLayout>
  );
};
