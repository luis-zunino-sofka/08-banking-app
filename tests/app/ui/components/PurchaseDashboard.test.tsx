import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import React, { act } from "react";
import { PurchaseDashboard } from "../../../../src/app/ui/components/purchaseDashboard/index";
import { decryptAES } from "../../../../src/app/core/utils/decryptAES.utils";
import { AccountContextMock } from "../containers/AppContainer.test";
import { mockAccountState, mockAppState } from "../../../mocks/dataMocked";

describe("PurchaseDashboard", () => {
  const mockOnSubmit = vi.fn();
  const mockRegister = vi.fn();
  const mockHandleSubmit = vi.fn((fn) => fn);
  // const mockRefetchAccounts = vi.fn();

  beforeEach(() => {
    render(
      <AccountContextMock
        mockAppState={mockAppState}
        children={<PurchaseDashboard />}
        mockAccountState={mockAccountState}
      />
    );

    vi.mock("'../../../../src/app/core/hooks/purchase/usePurchase'", () => ({
      usePurchase: () => ({
        isLoading: false,
        accounts: [
          {
            accountId: "1",
            encryptedNumber: `${decryptAES("encrypted1")}`,
            amount: 1000,
          },
          {
            accountId: "2",
            encryptedNumber: `${decryptAES("encrypted2")}`,
            amount: 2500,
          },
        ],
        errors: {
          purchaseType: { message: "Por favor selecciona un tipo de compra" },
          accountId: { message: "La cuenta es obligatoria" },
          purchaseAmount: { message: "El monto es obligatorio" },
        },
        register: mockRegister,
        handleSubmit: mockHandleSubmit,
        onSubmit: mockOnSubmit,
      }),
    }));
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  test("Match snapshoot", () => {
    const { asFragment } = render(
      <AccountContextMock
        mockAppState={mockAppState}
        children={<PurchaseDashboard />}
        mockAccountState={mockAccountState}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("debería renderizar el formulario correctamente", () => {
    screen.getByLabelText(/Selecciona el tipo de compra/i);
    screen.getByLabelText(/Selecciona la cuenta a depositar/i);
    screen.getByLabelText(/Ingresa el monto a depositar/i);
    screen.getByRole("button", { name: /Comprar/i });
  });

  test("debería mostrar error cuando falta seleccionar el tipo de compra", async () => {
    const submitButton = screen.getByRole("button", { name: /Comprar/i });
    const inputPurchaseType = screen.getByRole("combobox", {
      name: "Selecciona el tipo de compra:",
    });
    const inputAccountId = screen.getByRole("combobox", {
      name: "Selecciona la cuenta a depositar:",
    });
    const inputPurchaseAmount = screen.getByRole("spinbutton", {
      name: "Ingresa el monto a depositar:",
    });
    act(() => {
      fireEvent.submit(submitButton);
    });

    expect(inputPurchaseType).toHaveValue("");
    expect(inputAccountId).toHaveValue("");
    expect(inputPurchaseAmount).toHaveValue(null);
    await waitFor(() =>
      expect(
        screen.getByText(/Por favor selecciona un tipo de compra/i)
      ).toBeInTheDocument()
    );
  });

  test("debería mostrar las cuentas correctamente en el select", async () => {
    const accountSelect = screen.getByRole("combobox", {
      name: "Selecciona la cuenta a depositar:",
    });

    await waitFor(() => {
      expect(accountSelect).toHaveTextContent(
        "Elige una cuentaCuenta: encrypted1 - Monto: 1000Cuenta: encrypted2 - Monto: 2500"
      );
    });
  });

  test("debería llamar a onSubmit cuando se envía el formulario correctamente", async () => {
    const submitButton = screen.getByRole("button", { name: /Comprar/i });

    const inputPurchaseType = screen.getByRole("combobox", {
      name: "Selecciona el tipo de compra:",
    });
    const inputAccountId = screen.getByRole("combobox", {
      name: "Selecciona la cuenta a depositar:",
    });

    const inputPurchaseAmount = screen.getByRole("spinbutton", {
      name: /Ingresa el monto a depositar:/i,
    });

    waitFor(() => {
      fireEvent.change(inputPurchaseType, {
        target: { value: "online" },
      });

      fireEvent.change(inputAccountId, { target: { value: 1 } });

      fireEvent.change(inputPurchaseAmount, { target: { value: 100 } });

      fireEvent.click(submitButton);
    });

    expect(inputPurchaseType).toHaveValue("online");
    expect(inputPurchaseAmount).toHaveValue(100);
    expect(inputAccountId).toHaveValue("1");

    waitFor(() => {
      expect(mockOnSubmit).toHaveBeenCalled();
    });
  });
});
