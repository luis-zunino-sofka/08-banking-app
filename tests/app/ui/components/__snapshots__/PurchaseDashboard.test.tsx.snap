// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AccountProvider > Match snapshoot 1`] = `<DocumentFragment />`;

exports[`PurchaseDashboard > Match snapshoot 1`] = `
<DocumentFragment>
  <div>
    <form
      class="form"
      id="purchase-form"
    >
      <div>
        <label
          for="purchase-type"
        >
          Selecciona el tipo de compra:
        </label>
        <select
          id="purchase-type"
          name="purchaseType"
        >
          <option
            value=""
          >
            Elige el tipo de compra
          </option>
          <option
            value="physical"
          >
            Establecimiento físico (Sin costo)
          </option>
          <option
            value="online"
          >
            En línea (Costo de seguro de $5 USD)
          </option>
        </select>
      </div>
      <div>
        <label
          for="account-number-to-deposit"
        >
          Selecciona la cuenta a depositar:
        </label>
        <select
          id="account-number-to-deposit"
          name="accountId"
        >
          <option
            value=""
          >
            Elige una cuenta
          </option>
          <option
            value="1"
          >
            Cuenta: encrypted1 - Monto: 1000
          </option>
          <option
            value="2"
          >
            Cuenta: encrypted2 - Monto: 2500
          </option>
        </select>
      </div>
      <div>
        <label
          for="purchase-amount"
        >
          Ingresa el monto a depositar:
          <input
            id="purchase-amount"
            name="purchaseAmount"
            placeholder="Monto de la compra en $"
            type="number"
          />
        </label>
      </div>
      <button
        type="submit"
      >
        Comprar
      </button>
    </form>
    <div
      style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
    />
  </div>
</DocumentFragment>
`;
